### PR TITLE
Add validating webhook for shoots

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
This PR adds a validating webhook that is responsible for checking resources of the group `core.gardener.cloud` (at the moment only `shoot` resources).

With [GEP-4](https://github.com/gardener/gardener/blob/master/docs/proposals/04-new-core-gardener-cloud-apis.md) cloud specific information was externalized and as a consequence cannot be validated by the Gardener anymore. On the other hand, it should be prevented that invalid `shoot` objects are created or updated.

**Release note**:

```noteworthy operator
The `gardener-extension-provider-azure` now offers a validating webhook which checks shoot resources of type `azure` in the `core.gardener.cloud` group. Operators should register this webhook in the Garden cluster to further prevent invalid modifications on `azure` shoots.
```

### CODE REVIEW

File: charts/gardener-extension-validator-azure/charts/application/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart to deploy the gardener-extensions-validator-azure application related resources
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/Chart.yaml]:
```suggestion
description: A Helm chart to deploy the gardener-extension-validator-azure application related resources
```

File: charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
+gardener-extensions-validator-azure
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl]:
```suggestion
gardener-extension-validator-azure
```

Comment (rfranzke) [reply charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl]: FYI: We should also rename AWS...

File: charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    - CREATE
+    - UPDATE
+    resources:
+    - shoots
+  failurePolicy: Fail
+  objectSelector: {}
+  namespaceSelector: {}
+  clientConfig:
+    {{- if .Values.global.virtualGarden.enabled }}
+    url: {{ printf "https://gardener-extensions-validator-azure.%s/webhooks/validate-shoot" .Release.Namespace }}
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml]:
```suggestion
    url: {{ printf "https://gardener-extension-validator-azure.%s/webhooks/validate-shoot" .Release.Namespace }}
```

File: charts/gardener-extension-validator-azure/charts/runtime/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: A Helm chart to deploy the gardener-extensions-validator-azure runtime related resources
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/Chart.yaml]:
```suggestion
description: A Helm chart to deploy the gardener-extension-validator-azure runtime related resources
```

File: charts/gardener-extension-validator-azure/charts/application/values.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+global:
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/values.yaml]: This should be a symlink to `../../values.yaml`

File: charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl (hunk 2 above)
Comment (rfranzke) [reply charts/gardener-extension-validator-azure/charts/application/templates/_helpers.tpl]: But I guess this should be a symlink to `../../templates/helpers.tpl`?

File: charts/gardener-extension-validator-azure/charts/runtime/templates/_helpers.tpl [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/templates/_helpers.tpl]: symlink?

File: charts/gardener-extension-validator-azure/charts/runtime/templates/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+spec:
+  revisionHistoryLimit: 0
+  replicas: {{ .Values.global.replicaCount }}
+  selector:
+    matchLabels:
+{{ include "labels" . | indent 6 }}
+  template:
+    metadata:
+      annotations:
+        checksum/secret-gardener-extensions-validator-azure-cert: {{ include (print $.Template.BasePath "/secret-cert.yaml") . | sha256sum }}
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/templates/deployment.yaml]: s/extensions/extension everywhere

File: charts/gardener-extension-validator-azure/charts/runtime/templates/secret-cert.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-azure-cert
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/templates/secret-cert.yaml]:
```suggestion
  name: gardener-extension-validator-azure-cert
```

File: charts/gardener-extension-validator-azure/charts/runtime/templates/secret-kubeconfig.yaml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.global.kubeconfig }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-azure-kubeconfig
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/templates/secret-kubeconfig.yaml]:
```suggestion
  name: gardener-extension-validator-azure-kubeconfig
```

File: charts/gardener-extension-validator-azure/charts/runtime/values.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
+global:
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/runtime/values.yaml]: This should be a symlink to `../../values.yaml`

File: charts/gardener-extension-validator-azure/templates/_helpers.tpl [hunk 11]
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
+gardener-extensions-validator-azure
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/templates/_helpers.tpl]:
```suggestion
gardener-extension-validator-azure
```

File: cmd/gardener-extension-validator-azure/app/app.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+	controllercmd "github.com/gardener/gardener-extensions/pkg/controller/cmd"
+	"github.com/gardener/gardener/pkg/apis/core/install"
+	"github.com/spf13/cobra"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var log = logf.Log.WithName("gardener-extensions-validator-azure")
```
Comment (rfranzke) [inline cmd/gardener-extension-validator-azure/app/app.go]:
```suggestion
var log = logf.Log.WithName("gardener-extension-validator-azure")
```

File: example/40-validatingwebhookconfiguration.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: gardener-extensions-validator-azure
```
Comment (rfranzke) [inline example/40-validatingwebhookconfiguration.yaml]:
```suggestion
  name: gardener-extension-validator-azure
```

File: example/validator-azure-certs/tls.crt [hunk 14]
```diff
@@ -0,0 +1,23 @@
+-----BEGIN CERTIFICATE-----
```
Comment (rfranzke) [inline example/validator-azure-certs/tls.crt]: It doesn't harm, but this cert contains "Provider AWS" which might be confusing.

File: pkg/apis/azure/validation/infrastructure.go [hunk 15]
```diff
@@ -48,13 +48,10 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, resourc
 	// validation here will fail for those cases.
 	// TODO: remove the following block and uncomment below blocks once deployment into existing resource groups works properly.
 	if infra.ResourceGroup != nil {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("resourceGroup"), infra.ResourceGroup, "specifying an existing resource group is not supported yet"))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resourceGroup"), infra.ResourceGroup, "specifying an existing resource group is not supported yet"))
 	}
 
-	networksPath := field.NewPath("networks")
-	if len(infra.Networks.Workers) == 0 {
```
Comment (rfranzke) [inline pkg/apis/azure/validation/infrastructure.go]: Why was this removed?

File: pkg/apis/azure/validation/infrastructure.go [hunk 16]
```diff
@@ -67,8 +64,8 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, resourc
 		if infra.Networks.VNet.CIDR != nil {
 			allErrs = append(allErrs, field.Invalid(networksPath.Child("vnet", "cidr"), *infra.Networks.VNet.ResourceGroup, "specifying a cidr for an existing vnet is not possible"))
 		}
-		if *infra.Networks.VNet.ResourceGroup == *resourceGroupName {
-			allErrs = append(allErrs, field.Invalid(networksPath.Child("vnet", "resourceGroup"), *infra.Networks.VNet.ResourceGroup, "specifying an existing vnet is the cluster resource group is not supported"))
+		if infra.ResourceGroup != nil && *infra.Networks.VNet.ResourceGroup == infra.ResourceGroup.Name {
```
Comment (rfranzke) [inline pkg/apis/azure/validation/infrastructure.go]: Why have you added `infra.ResourceGroup != nil`? It's already in line 63

File: pkg/apis/azure/validation/infrastructure.go (hunk 15 above)
Comment (timuthy) [reply pkg/apis/azure/validation/infrastructure.go]: `cidrvalidation.NewCIDR` in line 56 already validates the correct CIDR range and an empty string is also invalid. We can also do it explicitly if you if you prefer the way it was done before?

Comment (timuthy) [reply pkg/apis/azure/validation/infrastructure.go]: The point is that before the user got two error messages:

```
              {
                  Type: "FieldValueRequired",
                  Field: "networks.workers",
                  BadValue: "",
                  Detail: "must specify the network range for the worker network",
              },
              {
                  Type: "FieldValueInvalid",
                  Field: "networks.workers",
                  BadValue: "",
                  Detail: "invalid CIDR address: ",
              },
```

So I thought it would be sufficient to only return `invalid CIDR address`.

**Comment by timuthy** [comment]
Thanks for the review @rfranzke. PTAL.

File: charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml [hunk 17]
```diff
@@ -0,0 +1,30 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: {{ include "name" . }}
+webhooks:
+- name: shoot.azure.provider.extensions.gardener.cloud
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml]:
```suggestion
- name: validation.azure.provider.extensions.gardener.cloud
```

File: charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+  objectSelector: {}
+  namespaceSelector: {}
+  clientConfig:
+    {{- if .Values.global.virtualGarden.enabled }}
+    url: {{ printf "https://%s.%s/webhooks/validate-shoot" (include "name" .) (.Release.Namespace) }}
+    {{- else }}
+    service:
+      namespace: {{ .Release.Namespace }}
+      name: {{ include "name" . }}
+      path: /webhooks/validate-shoot
```
Comment (rfranzke) [inline charts/gardener-extension-validator-azure/charts/application/templates/validatingwebhook-validator.yaml]:
```suggestion
      path: /webhooks/validate
```

File: example/40-validatingwebhookconfiguration.yaml [hunk 19]
```diff
@@ -0,0 +1,23 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: gardener-extension-validator-azure
+webhooks:
+- name: shoot.azure.provider.extensions.gardener.cloud
```
Comment (rfranzke) [inline example/40-validatingwebhookconfiguration.yaml]:
```suggestion
- name: validation.azure.provider.extensions.gardener.cloud
```

File: example/40-validatingwebhookconfiguration.yaml [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - shoots
+  failurePolicy: Fail
+  objectSelector: {}
+  namespaceSelector: {}
+  clientConfig:
+    url: "https://localhost:9443/webhooks/validate-shoot"
```
Comment (rfranzke) [inline example/40-validatingwebhookconfiguration.yaml]:
```suggestion
    url: "https://localhost:9443/webhooks/validate"
```

File: pkg/validator/shoot_validator.go [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+	networkPath := field.NewPath("spec", "networking")
+
+	if errList := azurevalidation.ValidateNetworking(shoot.Spec.Networking, networkPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Provider validation
+	fldPath := field.NewPath("spec", "provider")
+
+	if errList := azurevalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, fldPath); len(errList) != 0 {
```
Comment (zanetworker) [inline pkg/validator/shoot_validator.go]: shouldn't this be `infraConfigFldPath` instead? 

```suggestion
	if errList := azurevalidation.ValidateInfrastructureConfig(infraConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, infraConfigFldPath); len(errList) != 0 {
```

**Comment by timuthy** [comment]
I addressed the review comments in the latest update. Furthermore, @tim-ebert suggested to aggregate the errors instead of early exit everywhere. This suggestion has been adapted where applicable.

File: pkg/validator/shoot_validator.go [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+	allErrs = append(allErrs, azurevalidation.ValidateWorkersUpdate(oldShoot.Spec.Provider.Workers, shoot.Spec.Provider.Workers, workersPath)...)
+
+	allErrs = append(allErrs, v.validateShoot(shoot, infraConfig)...)
+
+	return allErrs.ToAggregate()
+}
+
+func (v *Shoot) validateShootCreation(shoot *core.Shoot) error {
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
```
Comment (rfranzke) [inline pkg/validator/shoot_validator.go]: Why are lines 93-100 not moved into the `validateShoot` function as well? The `ControlPlaneConfig` related code is in this function as well.

File: pkg/validator/shoot_validator.go [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+func (v *Shoot) validateShootUpdate(oldShoot, shoot *core.Shoot) error {
+	// InfrastructureConfig update
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(infraConfigPath, "InfrastructureConfig must be set for Azure shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig)
+	if err != nil {
+		return field.Forbidden(infraConfigPath, "not allowed to configure an unsupported infrastructureConfig")
+	}
```
Comment (rfranzke) [inline pkg/validator/shoot_validator.go]: Maybe it makes sense to combine this into a function as it is done multiple times (first `nil` check, then decode check).

File: pkg/validator/shoot_validator.go (hunk 22 above)
Comment (timuthy) [reply pkg/validator/shoot_validator.go]: You need to do it before because in `validateShootUpdate` also the "old" object needs to be decoded + checked. If it were in `validateShoot ` we'd do it twice.
